### PR TITLE
Feature/cloudstack affinity type

### PR DESCRIPTION
**Pull request by wongni**
*Issue #, if available:* #1601

*Description of changes:*
Support Affinity property for cloudstack provider

*Testing (if applicable):*
Manual testing for "anti", "pro" and "no" Affinity types by creating a management cluster and a workload cluster

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: config/manifest/eksa-components.yaml [hunk 1]
```diff
@@ -4792,7 +4796,7 @@ spec:
         - --health-probe-bind-address=:8081
         - --metrics-bind-address=127.0.0.1:8080
         - --leader-elect
-        image: public.ecr.aws/l0g8r8j6/eks-anywhere-cluster-controller:v0.0.0-eks-a-v0.0.0-dev-build.251
```
Comment (maxdrib) [inline config/manifest/eksa-components.yaml:4795]: please revert this change

File: controllers/controllers/resource/fetcher_test.go [hunk 2]
```diff
@@ -281,6 +282,7 @@ func TestMapMachineTemplateToCloudStackWorkerMachineConfigSpec(t *testing.T) {
 				Spec: anywherev1.CloudStackMachineConfigSpec{
 					Template:          anywherev1.CloudStackResourceIdentifier{Name: "rhel8-1.20"},
 					ComputeOffering:   anywherev1.CloudStackResourceIdentifier{Name: "large"},
+					Affinity:         "anti",
```
Comment (maxdrib) [inline controllers/controllers/resource/fetcher_test.go:285]: you can detect lint issues with `make lint` and then execute gofumpt on the folder/file to fix them

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go [hunk 3]
```diff
@@ -32,6 +32,8 @@ type CloudStackMachineConfigSpec struct {
 	ComputeOffering CloudStackResourceIdentifier `json:"computeOffering"`
 	// Users consists of an array of objects containing the username, as well as a list of their public keys. These users will be authorized to ssh into the machines
 	Users []UserConfiguration `json:"users,omitempty"`
+	// Defaults to `no`. Can be `pro` or `anti`. Will create an affinity group per machine set.
+	Affinity string `json:"affinity,omitempty"`
```
Comment (maxdrib) [inline pkg/api/v1alpha1/cloudstackmachineconfig_types.go:36]: Would this field be mutable, or only settable once? We should put a check for that in cloudstack.go under `AnyImmutableFieldChanged` and possibly the webhook if it's being modified illegally

Also, could you validate that the string is properly formatted under validator.go?

File: pkg/providers/cloudstack/config/template-cp.yaml [hunk 4]
```diff
@@ -435,6 +438,9 @@ spec:
         {{$key}}: {{$value}}
 {{- end}}
 {{- end}}
+{{- if .cloudstackEtcdAffinity }}
+      affinity: '{{ .cloudstackEtcdAffinity }}'
```
Comment (maxdrib) [inline pkg/providers/cloudstack/config/template-cp.yaml:442]: right now I see two very similar-sounding template substitution names: affinity and cloudstackAffinity. Could you make it more clear which one is for worker and which is for control plane?

File: pkg/api/v1alpha1/cloudstackmachineconfig_types.go [hunk 5]
```diff
@@ -32,6 +32,8 @@ type CloudStackMachineConfigSpec struct {
 	ComputeOffering CloudStackResourceIdentifier `json:"computeOffering"`
 	// Users consists of an array of objects containing the username, as well as a list of their public keys. These users will be authorized to ssh into the machines
 	Users []UserConfiguration `json:"users,omitempty"`
+	// Defaults to `no`. Can be `pro` or `anti`. Will create an affinity group per machine set.
```
Comment (maxdrib) [inline pkg/api/v1alpha1/cloudstackmachineconfig_types.go:35]: I think this description could use more detail. "If set, will create an affinity group per machine set of the corresponding type"

**Comment by chrisdoherty4** [comment]
@maxdrib Are we doing broader e2e testing for this kind of thing?

**Comment by maxdrib** [comment]
synced offline with @chrisdoherty4 and we feel like introducing a unit test for the template change would be appropriate here, since there's no actual business logic being added to eks-a, only plumbing work. However an e2e test for this feature in CAPC feels appropriate. Is that something we can introduce @wongni ?

File: controllers/controllers/resource/fetcher_test.go (hunk 2 above)
Comment (wongni) [reply controllers/controllers/resource/fetcher_test.go]: Fixed.

File: config/manifest/eksa-components.yaml (hunk 1 above)
Comment (wongni) [reply config/manifest/eksa-components.yaml:4795]: reverted

File: pkg/providers/cloudstack/config/template-cp.yaml (hunk 4 above)
Comment (wongni) [reply pkg/providers/cloudstack/config/template-cp.yaml:442]: Done

File: pkg/providers/cloudstack/cloudstack.go [hunk 6]
```diff
@@ -530,13 +531,14 @@ func AnyImmutableFieldChanged(oldCsdc, newCsdc *v1alpha1.CloudStackDatacenterCon
 	if oldCsmc.Spec.ComputeOffering != newCsmc.Spec.ComputeOffering {
 		return true
 	}
-	if len(oldCsmc.Spec.UserCustomDetails) != len(newCsmc.Spec.UserCustomDetails) {
+	if !reflect.DeepEqual(oldCsmc.Spec.UserCustomDetails, newCsmc.Spec.UserCustomDetails) {
```
Comment (maxdrib) [inline pkg/providers/cloudstack/cloudstack.go:534]: the general consensus is that we should avoid using reflect.DeepEqual outside of test code. Please revert this change

File: pkg/providers/cloudstack/cloudstack.go [hunk 7]
```diff
@@ -530,13 +531,14 @@ func AnyImmutableFieldChanged(oldCsdc, newCsdc *v1alpha1.CloudStackDatacenterCon
...
 	}
-	for key, value := range oldCsmc.Spec.UserCustomDetails {
-		if value != newCsmc.Spec.UserCustomDetails[key] {
-			return true
-		}
+	if oldCsmc.Spec.Affinity != newCsmc.Spec.Affinity {
+		return true
+	}
+	if !reflect.DeepEqual(oldCsmc.Spec.AffinityGroupIds, newCsmc.Spec.AffinityGroupIds) {
+		return true
```
Comment (maxdrib) [inline pkg/providers/cloudstack/cloudstack.go:541]: can capc handle changing these attributes?

Comment (wongni) [reply pkg/providers/cloudstack/cloudstack.go]: No, capc doesn't allow them to be changed.

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go [hunk 8]
```diff
@@ -47,7 +48,46 @@ func (r *CloudStackMachineConfig) ValidateUpdate(old runtime.Object) error {
...
+	var allErrs field.ErrorList
+
+	if old.Spec.Affinity != new.Spec.Affinity {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "affinity"), new.Spec.Affinity, "field is immutable"),
+		)
+	}
+
+	afffinityGroupIdsMutated := false
```
Comment (maxdrib) [inline pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:72]: typo: too many f's

File: pkg/providers/cloudstack/cloudstack.go [hunk 9]
```diff
@@ -636,12 +636,14 @@ func buildTemplateMapCP(clusterSpec *cluster.Spec, datacenterConfigSpec v1alpha1
 		"cloudstackControlPlaneTemplateOfferingId":   controlPlaneMachineSpec.Template.Id,
 		"cloudstackControlPlaneTemplateOfferingName": controlPlaneMachineSpec.Template.Name,
 		"cloudstackControlPlaneCustomDetails":        controlPlaneMachineSpec.UserCustomDetails,
-		"affinityGroupIds":                           controlPlaneMachineSpec.AffinityGroupIds,
```
Comment (maxdrib) [inline pkg/providers/cloudstack/cloudstack.go:639]: nice catch & thanks for fixing

File: pkg/providers/cloudstack/config/template-cp.yaml (hunk 4 above)
Comment (maxdrib) [inline pkg/providers/cloudstack/config/template-cp.yaml:442]: I'm not sure if we need the quotes around the value here

**Review by maxdrib — COMMENTED** [review]
Looks like unit tests are failing. Nice work, thanks for driving this change

File: pkg/providers/cloudstack/config/template-cp.yaml (hunk 4 above)
Comment (wongni) [reply pkg/providers/cloudstack/config/template-cp.yaml:442]: 'no' value without quotes is treated as a boolean value so they are needed.

File: pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go (hunk 8 above)
Comment (wongni) [reply pkg/api/v1alpha1/cloudstackmachineconfig_webhook.go:72]: Done

**Comment by wongni** [comment]
Created a new PR on the top of the main branch: https://github.com/aws/eks-anywhere/pull/1682
